### PR TITLE
[DO NOT MERGE]CHAT-2219: Modify libjingle to support BOSH in Webrtc XMPP stack

### PR DESCRIPTION
**Pull request by DinhNguyenCat**
//cc: @dscherba @digitalmouse12 @willkelleher 

introduce a singleton object simplebosh to wrap all outgoing xmpp xml message into a suitable http request.
This object also parse the incoming http message to get the xml content to process.
Two class util and tag is used to support simplebosh object to do its functionality.

### CODE REVIEW

**Comment by dscherba** [comment]
:-1:, see non-public comments [here](https://personify.atlassian.net/browse/CHAT-2219)

**Comment by DinhNguyenCat** [comment]
//cc: @dscherba @digitalmouse12 @willkelleher 

update the code with some improvement:
Remove singleton pattern.
Flexible behavior based on port number, if we open socket with port 5222, it will send normal xml message, if we open socket with port 80, it will send out BOSH message.
Rename the singlebosh class in to ContentWrapper class. The new class has the same functionality with the original class --> This class is used to wrap xml message in normal case (XMPP message)
Introduce new class BoshContentWrapper, inherited from ContentWrapper, used to wrap xml message in BOSH format.
Remove source code of gloox library.

**Comment by dscherba** [comment]
@DinhNguyenCat : will take another look. 

In the meantime, see https://github.com/PersonifyInc/StagePresence/pull/574 (we're getting very close to merging this WebRTC update). You'll want to update your work to be compatible, which should be straight-forward, but may need some namespace changes.

**Comment by DinhNguyenCat** [comment]
@dscherba I will check again the new update and update my work accordingly. Is this the new update from Google's WebRTC lib? And when will we merge the update into our code (in dependencies)?

**Comment by dscherba** [comment]
The new update is a drop of WebRTC from upstream (and associated stuff to
make it work for our needs). Looking to get it merged in the next 24 hours,
or so, given that some references in our previous drop have gone stale.

Look at that PR, specifically the .../webrtc2/README.md file for
gclient/build invocations.

I don't expect that the XMPP layer will be much different.

File: talk/xmpp/contentwrapper.cc [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    requestbody += "xmlns=\'" + kXmlnsHttpbind + "\'/>";
+    rid = 0;
+    sid = "";
+    start_sent = false;
+    return requestbody;
+  }
+
+  std::string BoshContentWrapper::GetStartString()
+  {
+    rid = rand() % 100000 + 1728679472;
```
Comment (dscherba) [inline talk/xmpp/contentwrapper.cc:100]: Why this? Seems like a 'magic' number (the 1728679472) and I don't believe this is speced

File: talk/xmpp/xmppclient.cc [hunk 2]
```diff
@@ -154,8 +154,10 @@ XmppReturnStatus XmppClient::Connect(
 
   // Set language
   d_->engine_->SetLanguage(lang);
-
```
Comment (dscherba) [inline talk/xmpp/xmppclient.cc:157]: let's avoid spacing changes for now (clutters the diff)

File: talk/xmpp/xmppengineimpl.cc [hunk 3]
```diff
@@ -104,10 +105,12 @@ XmppReturnStatus XmppEngineImpl::HandleInput(
     return XMPP_RETURN_BADSTATE;
 
   EnterExit ee(this);
-
+  std::string str(bytes);
+  if(!content_wrapper_->HandleRecvData(str, len))
+    return XMPP_RETURN_UNEXPECTED;
```
Comment (dscherba) [inline talk/xmpp/xmppengineimpl.cc:110]: Would prefer that we always use braces (even for one-liners)

File: talk/xmpp/xmppengineimpl.h [hunk 4]
```diff
@@ -102,6 +103,27 @@ class XmppEngineImpl : public XmppEngine {
...
+    connection_port_ = port;
+
+    std::string hostname = tls_server_hostname_;
+    if (hostname.empty())
+      hostname = this->user_jid_.domain();
+
+    // If not language is specified, the spec says use *
+    std::string lang = lang_;
+    if (lang.length() == 0)
+      lang = "*";
```
Comment (dscherba) [inline talk/xmpp/xmppengineimpl.h:118]: braces

File: talk/xmpp/xmppengineimpl.h [hunk 5]
```diff
@@ -102,6 +103,27 @@ class XmppEngineImpl : public XmppEngine {
...
+
+    // Hard code the port number here, need improvement?
+    if (port == 5222)
+      content_wrapper_ = new ContentWrapper(hostname, lang);
```
Comment (dscherba) [inline talk/xmpp/xmppengineimpl.h:122]: braces (always, please). Won't try to tag other instances, but check.

**Comment by dscherba** [comment]
@DinhNguyenCat : I took some time to build this PR locally and was eventually able to run it. Here's my next round of feedback: 
1. As noted before, you'll want to rebase/adjust your changes to work with the updated WebRTC. This should make things easier, long-term, since the new may is more amenable to changing WebRTC (i.e. we don't have a cached, versioned WebRTC binary)
2. This presently doesn't work with PRODUCT builds due to some code in Xmpp/xmpp_client.cpp that sets the buzz::TLS_REQUIRED flag for PRODUCT builds. The BOSH code doesn't work with this flag yet.
3. Made some style comments in the commit (bracing all ifs, avoiding whitespace changes)
4. This implementation doesn't use HTTP 1.1 keep-alive...
5. ... but it's possible for this to send multiple HTTP requests before all responses are received (termed 'HTTP pipelining')--this is invalid HTTP behavior and I suspect this might cause some issues.
6. ... and this doesn't have the long-polling behavior that we're looking for with BOSH. Reference http://xmpp.org/extensions/xep-0124.html for a detailed discussion, but with this implementation the client will only get asynchronous notifications from the server in some future response. This isn't completely awful since the client pings the server regularly, but it can cause delays in many use cases: getting invites, text messages, etc...

I think we should address the above before we start worrying about running this over a proxy.

File: talk/xmpp/contentwrapper.cc (hunk 1 above)
Comment (DinhNguyenCat) [inline talk/xmpp/contentwrapper.cc:100]: Noted, will update

File: talk/xmpp/xmppclient.cc (hunk 2 above)
Comment (DinhNguyenCat) [inline talk/xmpp/xmppclient.cc:157]: Noted

File: talk/xmpp/xmppengineimpl.cc (hunk 3 above)
Comment (DinhNguyenCat) [inline talk/xmpp/xmppengineimpl.cc:110]: Noted

File: talk/xmpp/xmppengineimpl.h (hunk 4 above)
Comment (DinhNguyenCat) [inline talk/xmpp/xmppengineimpl.h:118]: Hi @dscherba,

In the original code of WebRTC, I saw that they usually didn't use braces in one line if command. You can easily see it in the code.

File: talk/xmpp/xmppengineimpl.h (hunk 5 above)
Comment (DinhNguyenCat) [inline talk/xmpp/xmppengineimpl.h:122]: Noted.

**Comment by DinhNguyenCat** [comment]
@dscherba  Thank you so much for your feedback.
1. I'm trying to rebase code with the new update, the network is quite slow so I have to take sometimes to finish
2. I will check again with TLS_REQUIRED= true. These days I only check with TLS disabled
3. Noted.
4. 5. HTTP 1.1 is considered keep alive unless declared otherwise.
6. It's needed to take sometimes to investigate the http long-polling connection. Currently I reused the asynchronous connection. Sorry for that. I will continue checking this problem.

**Comment by DinhNguyenCat** [comment]
Close this PR since it's moved to https://github.com/PersonifyInc/webrtc/pull/4

**Pull request closed without merging**